crates/testing/src/block_builder.rs

```diff
@@ -1,36 +1,151 @@
 use tide_disco::{method::ReadState, App, Url};
 
-/// The only block [`TestableBuilderSource`] provides
-const EMPTY_BLOCK: TestBlockPayload = TestBlockPayload {
-    transactions: vec![],
-};
+use crate::test_runner::HotShotTaskCompleted;
+
+/// Entry for a built block
+struct BlockEntry {
+    metadata: AvailableBlockInfo<TestTypes>,
+    payload: Option<AvailableBlockData<TestTypes>>,
+    header_input: Option<AvailableBlockHeaderInput<TestTypes>>,
+}
+
+/// Options controlling how the random builder generates blocks
+#[derive(Clone, Debug)]
+pub struct RandomBuilderOptions {
+    /// How many transactions to include in a block
+    pub txn_in_block: u64,
+    /// How many blocks to generate per second
+    pub blocks_per_second: u64,
+    /// Range of how big a transaction can be (in bytes)
+    pub txn_size: Range<u32>,
+    /// Number of storage nodes for VID commitment
+    pub num_storage_nodes: usize,
+}
+
+impl Default for RandomBuilderOptions {
+    fn default() -> Self {
+        Self {
+            txn_in_block: 100,
+            blocks_per_second: 1,
+            txn_size: 20..100,
+            num_storage_nodes: 1,
+        }
+    }
+}
 
 /// A mock implementation of the builder data source.
-/// "Builds" only empty blocks.
-pub struct TestableBuilderSource {
-    priv_key: <<TestTypes as NodeType>::SignatureKey as SignatureKey>::PrivateKey,
+/// Builds random blocks, doesn't track HotShot state at all.
+/// Evicts old available blocks if HotShot doesn't keep up.
+#[derive(Clone, Debug)]
+pub struct RandomBuilderSource {
+    /// Built blocks
+    blocks: Arc<
+        RwLock<
+            // Isn't strictly speaking used as a cache,
+            // just as a HashMap that evicts old blocks
+            LruCache<BuilderCommitment, BlockEntry>,
+        >,
+    >,
     pub_key: <TestTypes as NodeType>::SignatureKey,
+    priv_key: <<TestTypes as NodeType>::SignatureKey as SignatureKey>::PrivateKey,
+}
+
+impl RandomBuilderSource {
+    /// Create new [`RandomBuilderSource`]
+    #[must_use]
+    #[allow(clippy::missing_panics_doc)] // ony panics if 256 == 0
+    pub fn new(
+        pub_key: <TestTypes as NodeType>::SignatureKey,
+        priv_key: <<TestTypes as NodeType>::SignatureKey as SignatureKey>::PrivateKey,
+    ) -> Self {
+        Self {
+            blocks: Arc::new(RwLock::new(LruCache::new(NonZeroUsize::new(256).unwrap()))),
+            priv_key,
+            pub_key,
+        }
+    }
+
+    /// Spawn a task building blocks, configured with given options
+    #[allow(clippy::missing_panics_doc)] // ony panics on 16-bit platforms
+    pub fn run(&self, options: RandomBuilderOptions) {
+        let blocks = self.blocks.clone();
+        let (priv_key, pub_key) = (self.priv_key.clone(), self.pub_key);
+        async_spawn(async move {
+            let mut rng = SmallRng::from_entropy();
+            let time_per_block = Duration::from_millis(1000 / options.blocks_per_second);
```

Comment:
The random number here is a bit hard to parse. Could it perhaps be a named constant? For example, is 1000 the block epoch time? It's hard to know in this case.

Reply:
Right, I already have no idea what this is about either :)

Reply:
I re-wrote it as `Duration::from_secs(1) / options.blocks_per_second`, that should make it clear hopefully.